### ISSUE TITLE
XDMP-NODB error when DB exists

### ISSUE THREAD

**Issue by damonfeldman** [marklogic-community/ml-gradle#291]
Not a gradle bug, but logging for visibility (and possible workarounds).

If ML has an amp that refers to a non-existent database, then the GET call to /manage/v2/amps fails with
```
ERROR Logging HTTP response body to assist with debugging: <error xmlns="http://marklogic.com/xdmp/error">
  <status-code>500</status-code>
  <status>Internal Server Error</status>
  <message-code>XDMP-NODB</message-code>
  <message>XDMP-NODB: xdmp:database-name(xs:unsignedLong("2284049253264589806")) -- No database with identifier 2284049253264589806</message>
</error>
```

## Workaround
The workaround is to find the amp in the admin console (port 8001) and remove it. This is a known bug in certain versions of MarkLogic, and will be fixed in the core product.

An ambitious ml-gradle contributor could catch it, detect it, and explain differently (e.g. by calling other status functions and finding the offending amp, if any)

**Comment by tomaszs86** [comment]
It's common bug. Sometimes I need to remove amp and sometime just a forest with incorrect database name from admin console (8001). This error is misleading because it blocks deployment using mlDeploy command.

**Comment by derms** [comment]
FYI this is a temporary workaround for the problem that you can put in your gradle script

It works (even if it is a bit ugly)

```
task mlRemoveOrphanedAmps(type:com.marklogic.gradle.task.MarkLogicTask) {
    doLast {
        def stringToEval = '''
      xquery version "1.0-ml";
      let $s :=
      "xquery version '1.0-ml';
        import module namespace sec='http://marklogic.com/xdmp/security' at '/MarkLogic/security.xqy';
        import module namespace functx = 'http://www.functx.com' at '/MarkLogic/functx/functx-1.0-nodoc-2007-01.xqy';
        let $dbs := xdmp:databases()
        for $amp in fn:collection(sec:amps-collection())/sec:amp[sec:database/text()!=0]
          where not(functx:is-value-in-sequence(xs:unsignedLong($amp/sec:database/text()),$dbs))
          return 'Removed ' || ($amp/sec:local-name,sec:remove-amp($amp/sec:namespace,$amp/sec:local-name,$amp/sec:document-uri,$amp/sec:database))"

        return xdmp:eval($s, (),
                  <options xmlns="xdmp:eval">
                    <database>{xdmp:database("Security")}</database>
                  </options>)
        '''
        def map = new org.springframework.util.LinkedMultiValueMap();
        map.add("xquery", stringToEval);
        def headers = new org.springframework.http.HttpHeaders();
        headers.setContentType(org.springframework.http.MediaType.APPLICATION_FORM_URLENCODED);
        def entity = new org.springframework.http.HttpEntity(map, headers);

        def uri = manageClient.buildUri("/v1/eval")
        println manageClient.adminRestTemplate.exchange(uri, org.springframework.http.HttpMethod.POST, entity, String.class);
    }
}

mlDeployApp.dependsOn(mlRemoveOrphanedAmps)
mlPostUndeploy.dependsOn(mlRemoveOrphanedAmps)
```

**Comment by derms** [comment]
I've also had an issue with undeploying json defined amps.

This is a temporary workaround for the undeployment problem also

```
task mlRemoveAmps(type:com.marklogic.gradle.task.MarkLogicTask) {
    doLast {
        appConfig.configDirs.each{cfg ->  cfg.ampsDir.eachFileMatch(FileType.FILES,~/.*.json/) { file -> 
            def amp =  new groovy.json.JsonSlurper().parseText(file.text)
            def localName = amp.'local-name'
            def documentUri = amp.'document-uri'
            def namespace = amp.'namespace' == null ? '' : amp.'namespace'
            def modulesDb = mlModulesDbName //could take from amp file or from another gradle property (e.g. mlModulesDbName)
            def uri = "/manage/v2/amps/$localName?namespace=$namespace&document-uri=$documentUri&modules-database=$modulesDb"             
            println "Deleting amp at '$uri'"
            manageClient.deleteAsAdmin(uri)
        }}
    }
}

task mlRemoveOrphanedAmps(type:com.marklogic.gradle.task.MarkLogicTask) {
    doLast {
        def stringToEval = '''
      xquery version "1.0-ml";
      let $s :=
      "xquery version '1.0-ml';
        import module namespace sec='http://marklogic.com/xdmp/security' at '/MarkLogic/security.xqy';
        import module namespace functx = 'http://www.functx.com' at '/MarkLogic/functx/functx-1.0-nodoc-2007-01.xqy';
        let $dbs := xdmp:databases()
        for $amp in fn:collection(sec:amps-collection())/sec:amp[sec:database/text()!=0]
          where not(functx:is-value-in-sequence(xs:unsignedLong($amp/sec:database/text()),$dbs))
          return 'Removed ' || ($amp/sec:local-name,sec:remove-amp($amp/sec:namespace,$amp/sec:local-name,$amp/sec:document-uri,$amp/sec:database))"

        return xdmp:eval($s, (),
                  <options xmlns="xdmp:eval">
                    <database>{xdmp:database("Security")}</database>
                  </options>)
        '''
        def map = new org.springframework.util.LinkedMultiValueMap();
        map.add("xquery", stringToEval);
        def headers = new org.springframework.http.HttpHeaders();
        headers.setContentType(org.springframework.http.MediaType.APPLICATION_FORM_URLENCODED);
        def entity = new org.springframework.http.HttpEntity(map, headers);

        def uri = manageClient.buildUri("/v1/eval")
        println manageClient.adminRestTemplate.exchange(uri, org.springframework.http.HttpMethod.POST, entity, String.class);
    }
}

mlUndeployAmps.dependsOn(mlRemoveAmps)
mlDeployApp.dependsOn(mlRemoveOrphanedAmps,mlRemoveAmps)
mlPostUndeploy.dependsOn(mlRemoveOrphanedAmps,mlRemoveAmps)
```

**Comment by dmcassel** [comment]
For internal reference, it's bugtrack #48844

**Comment by rjrudin** [comment]
48844 was fixed in 9.0-9, so closing this.
